src/Microsoft.IdentityModel.Clients.ActiveDirectory/Internal/Instance/InstanceDiscovery.cs

```diff
@@ -187,6 +187,11 @@ private async Task DiscoverAsync(Uri authority, bool validateAuthority, RequestC
                 // All other unexpected exceptions will still bubble up, as always.
                 if (validateAuthority)
                 {
+                    if (ex.ErrorCode == ErrorCodes.UnknownError && ex.StatusCode == 502)
```

Comment:
This is a bit tricky because HTTP error 502 is not the only one that represents a timeout. There are several others that represent the same thing. 

In my mind, we should just re-throw all exceptions except of when ex.ErrorCode == "invalid_instance". Neither ADAL nor MSAL will work without instance discovery anyway, so we are not introducing a breaking change.

So I would do smth like 

if error is "invalid_instance" throw AdalError.AuthorityNotInValidList
else rethrow